### PR TITLE
Store detector DLLs on disk

### PR DESCRIPTION
**Pull request by liadtal**

### CODE REVIEW

File: src/SDK/MonitoringApplianceEmulator/App.xaml.cs [hunk 1]
```diff
@@ -25,6 +27,8 @@ namespace Microsoft.Azure.Monitoring.SmartDetectors.MonitoringAppliancEmulator
     /// </summary>
     public partial class App : Application
     {
+        private static string tempFolder;
```
Comment (yanivid3) [inline src/SDK/MonitoringApplianceEmulator/App.xaml.cs]: >tempFolder [](start = 30, length = 10)

Dont we have a convesion for members?
_tempFolder

File: src/SDK/MonitoringApplianceEmulator/App.xaml.cs [hunk 2]
```diff
@@ -92,6 +101,36 @@ protected override void OnStartup(StartupEventArgs e)
                 .RegisterInstance(stateRepositoryFactory);
         }
 
+        /// <summary>
+        /// Raises the <see cref="E:System.Windows.Application.Exit" /> event.
+        /// </summary>
+        /// <param name="e">An <see cref="T:System.Windows.ExitEventArgs" /> that contains the event data.</param>
+        protected override void OnExit(ExitEventArgs e)
```
Comment (yanivid3) [inline src/SDK/MonitoringApplianceEmulator/App.xaml.cs]: >OnExit [](start = 32, length = 6)

why?
Don't we want these DLLs for next emulator run?

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs [hunk 3]
```diff
@@ -53,6 +51,11 @@ public static string Hash(this string s)
         /// <returns>The evaluated result string</returns>
         public static string EvaluateInterpolatedString(string interpolatedStringDefinition, object source)
         {
+            if (!interpolatedStringDefinition.Contains("{"))
```
Comment (yanivid3) [inline src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: >interpolatedStringDefinition [](start = 17, length = 28)

why? did it throw in such cases?

**Comment by yanivid3** [comment]
>     /// Implementation of the <see cref="ISmartDetectorLoader"/> interface.

I hate these documentations! They add nothing, and I can't understand what this class is about! :@:@:@

---
Refers to: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs:18 in 071e23e. [](commit_id = 071e23eade70d3d43f62fdbb56636a82a78b6d48, deletion_comment = False)

File: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs [hunk 4]
```diff
@@ -50,33 +54,27 @@ public ISmartDetector LoadSmartDetector(SmartDetectorPackage smartDetectorPackag
             {
                 this.tracer.TraceInformation($"Read {smartDetectorPackage.Content.Count} assemblies for Smart Detector ID {smartDetectorManifest.Id}");
 
+                // Write all DLLs to the temp folder
+                foreach (var assemblyNameAndBytes in smartDetectorPackage.Content)
+                {
+                    string fileName = Path.Combine(this.tempFolder, assemblyNameAndBytes.Key);
+                    File.WriteAllBytes(fileName, assemblyNameAndBytes.Value);
```
Comment (yanivid3) [inline src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs]: >File.WriteAllBytes [](start = 20, length = 18)

what happens if the file exists? is it always overridden? (Needed in case of a newer version)

**Comment by yanivid3** [comment]
> namespace Microsoft.Azure.Monitoring.SmartDetectors.SmartDetectorLoader

do you remember the issue we had here with tne Math nuget I tried to add?
Can you please make sure it works after the changes you made here ?

---
Refers to: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs:7 in 071e23e. [](commit_id = 071e23eade70d3d43f62fdbb56636a82a78b6d48, deletion_comment = False)

File: test/SmartDetectorsSharedTests/SmartDetectorLoaderTests.cs [hunk 5]
```diff
@@ -32,14 +32,46 @@ public class SmartDetectorLoaderTests
         private Mock<ITracer> tracerMock;
         private Dictionary<string, SmartDetectorManifest> manifests;
         private Dictionary<string, Dictionary<string, byte[]>> assemblies;
+        private string tempFolder;
+
+        public TestContext TestContext { get; set; }
 
         [TestInitialize]
         public void TestInitialize()
```
Comment (yanivid3) [inline test/SmartDetectorsSharedTests/SmartDetectorLoaderTests.cs]: >TestInitialize [](start = 20, length = 14)

no new tests are needed?

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs [hunk 6]
```diff
@@ -98,6 +99,19 @@ public ITracer CreateTracerForChildProcess(string[] args)
             return tracerForChildProcess;
         }
 
+        /// <summary>
+        /// Creates a smart detector loader object for the child process, based on the command line arguments received from the parent process.
```
Comment (yanivid3) [inline src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: >based on the command line arguments [](start = 74, length = 35)

1. is it documented that the value is expected in 4th arg?
2. That's the only arg you are using, can't you set the signature to:
CreateLoaderForChildProcess(string path, ITracer tracerForLoader)
?

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs [hunk 7]
```diff
@@ -324,6 +355,59 @@ public ITracer CreateTracerForChildProcess(string[] args)
...
+        private void CleanupTempFolder(string tempFolderPath)
+        {
+            DirectoryInfo tempFolder = new DirectoryInfo(tempFolderPath);
+
+            // Delete the specified temp folder
+            this.TryDeleteFolder(tempFolder);
+
+            // Delete the all folders in the parent of the temp folder that were created over 6 hours ago
+            // This handles scenarios of a previous crash that prevented the folder from being deleted
+            tempFolder.Parent?.GetDirectories()
```
Comment (yanivid3) [inline src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: >GetDirectories [](start = 31, length = 14)

if we want to delete - why not do it in emulator as well?

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs [hunk 8]
```diff
@@ -324,6 +355,59 @@ public ITracer CreateTracerForChildProcess(string[] args)
...
+            }
+
+            return tempFolder;
+        }
+
+        /// <summary>
+        /// Cleanup the temp folder
+        /// </summary>
+        /// <param name="tempFolderPath">The temp folder used for the child process (created by <see cref="CreateTempFolder"/>)</param>
+        private void CleanupTempFolder(string tempFolderPath)
```
Comment (yanivid3) [inline src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: >CleanupTempFolder [](start = 21, length = 17)

I suggest to reuse these funcs in emulator to make the code more maintainable

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs [hunk 9]
```diff
@@ -324,6 +355,59 @@ public ITracer CreateTracerForChildProcess(string[] args)
...
+        /// Try to delete the specified folder - does not throw exception in case of failure
+        /// </summary>
+        /// <param name="folder">The folder to delete</param>
+        /// <returns>True if the folder was successfully deleted, false otherwise</returns>
+        private bool TryDeleteFolder(DirectoryInfo folder)
+        {
+            try
+            {
+                folder.Delete(true);
+                this.tracer.TraceWarning($"Successfully deleted temp folder {folder.FullName}");
```
Comment (ilaizi) [inline src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: >TraceWarning [](start = 28, length = 12)

Should this be TraceInformation?

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs [hunk 10]
```diff
@@ -324,6 +355,59 @@ public ITracer CreateTracerForChildProcess(string[] args)
...
+                return true;
+            }
+            catch (Exception e)
+            {
+                this.tracer.TraceWarning($"Could not delete temp folder {folder.FullName}, error: {e.Message}");
```
Comment (yanivid3) [inline src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: >this.tracer.TraceWarning($"Could not delete temp folder {folder.FullName}, error: {e.Message}"); [](start = 16, length = 96)

please report an exception so we will see it. this can cause issues if deletes fail a lot

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs [hunk 11]
```diff
@@ -168,12 +186,21 @@ public ITracer CreateTracerForChildProcess(string[] args)
...
                             string customPropertiesJson = JsonConvert.SerializeObject(customProperties);
-                            string customPropertiesString = Convert.ToBase64String(Encoding.UTF8.GetBytes(customPropertiesJson));
+
+                            // Prepare command line arguments
+                            string[] arguments = new string[]
+                            {
+                                pipeParentToChildHandle,
+                                pipeChildToParentHandle,
+                                this.tracer.SessionId,
+                                Convert.ToBase64String(Encoding.UTF8.GetBytes(customPropertiesJson)),
```
Comment (yanivid3) [inline src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: >Convert.ToBase64String(Encoding.UTF8.GetBytes [](start = 32, length = 45)

WHy did you change it 
"Convert.ToBase64String(Encoding.UTF8.GetBytes"?

**Comment by EranGon** [comment]
>         /// Registers a tracer instance for the child process, based on the arguments received from the parent process.

fix

---
Refers to: src/analysis/SmartDetectorsAnalysis/AnalysisDependenciesInjector.cs:52 in 071e23e. [](commit_id = 071e23eade70d3d43f62fdbb56636a82a78b6d48, deletion_comment = False)

File: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs [hunk 12]
```diff
@@ -50,33 +54,27 @@ public ISmartDetector LoadSmartDetector(SmartDetectorPackage smartDetectorPackag
...
-
-                    // Try to find the assembly bytes in the Smart Detector's assemblies
-                    if (smartDetectorPackage.TryGetAssemblyBytes(name, out byte[] assemblyBytes))
-                    {
-                        // Load the assembly from its bytes
-                        return Assembly.Load(assemblyBytes);
-                    }
-
-                    return null;
+                    return this.LoadAssembly(args.Name);
```
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs]: >return this.LoadAssembly(args.Name); [](start = 20, length = 36)

why do you need this at all?

File: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs [hunk 13]
```diff
@@ -135,6 +133,39 @@ public ISmartDetector LoadSmartDetector(SmartDetectorPackage smartDetectorPackag
             }
         }
 
+        /// <summary>
+        /// Load an assembly from the temporary folder
+        /// </summary>
+        /// <param name="assemblyNameString">The assembly name</param>
+        /// <returns>The loaded assembly</returns>
+        private Assembly LoadAssembly(string assemblyNameString)
```
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs]: >LoadAssembly [](start = 25, length = 12)

i think this is the default behaviour of Assembly

**Comment by EranGon** [comment]
>         public ISmartDetector LoadSmartDetector(SmartDetectorPackage smartDetectorPackage)

suggestion:
- change the method to return a new class: LoadedSmartDetector (or some better name hopefully)
- have the class encapsulate both the detector and the path where it was loaded from
- make the class IDisposable, and delete the folder on Dispose
 

---
Refers to: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs:49 in 071e23e. [](commit_id = 071e23eade70d3d43f62fdbb56636a82a78b6d48, deletion_comment = False)

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs [hunk 14]
```diff
@@ -98,6 +99,19 @@ public ITracer CreateTracerForChildProcess(string[] args)
...
+        /// <summary>
+        /// Creates a smart detector loader object for the child process, based on the command line arguments received from the parent process.
+        /// </summary>
+        /// <param name="args">The command line arguments.</param>
+        /// <param name="tracerForLoader">The tracer to use</param>
+        /// <returns>The smart detector loader instance</returns>
+        public ISmartDetectorLoader CreateLoaderForChildProcess(string[] args, ITracer tracerForLoader)
+        {
+            // Get the temp folder name from the arguments and create the loader
+            string tempFolder = Encoding.UTF8.GetString(Convert.FromBase64String(args[4]));
```
[the remark this answers is not in the document]
Comment (EranGon) [reply src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: oh, because you can't delete the folder from the child process since you can't unload the assembly?

---
In reply to: [186694207](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186694207) [](ancestors = 186694207)

File: src/SDK/MonitoringApplianceEmulator/App.xaml.cs (hunk 1 above)
Comment (liadtal) [reply src/SDK/MonitoringApplianceEmulator/App.xaml.cs]: Nope - tempFolder is static, no need for "this"

---
In reply to: [186694967](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186694967) [](ancestors = 186694967,186693426,186682735)

File: src/SDK/MonitoringApplianceEmulator/App.xaml.cs (hunk 2 above)
Comment (liadtal) [reply src/SDK/MonitoringApplianceEmulator/App.xaml.cs]: Then we will extract them again from the package

---
In reply to: [186683869](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186683869) [](ancestors = 186683869)

**Comment by liadtal** [comment]
> namespace Microsoft.Azure.Monitoring.SmartDetectors.SmartDetectorLoader

Yes, it was tested with dependent DLLs

---
In reply to: [387362873](https://github.com/liadtal/Smart-Alerts/pull/150#issuecomment-387362873) [](ancestors = 387362873)

---
Refers to: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs:7 in 071e23e. [](commit_id = 071e23eade70d3d43f62fdbb56636a82a78b6d48, deletion_comment = False)

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (liadtal) [reply src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: Exactly. The parent process deletes this folder, after the child process completes

---
In reply to: [186696031](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186696031) [](ancestors = 186696031,186694207)

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs (hunk 9 above)
Comment (liadtal) [reply src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: thanks...

---
In reply to: [186689337](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186689337) [](ancestors = 186689337)

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs (hunk 10 above)
Comment (liadtal) [reply src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: Thanks, good point

---
In reply to: [186689392](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186689392) [](ancestors = 186689392)

File: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs (hunk 13 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs]: The default behavior is to search for the assembly in the current executable's folder - not in the temp folder. That's why we need this method.

---
In reply to: [186693331](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186693331) [](ancestors = 186693331)

File: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs [hunk 15]
```diff
@@ -18,14 +19,17 @@ namespace Microsoft.Azure.Monitoring.SmartDetectors.SmartDetectorLoader
...
+        private readonly string tempFolder;
         private readonly ITracer tracer;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="SmartDetectorLoader"/> class.
         /// </summary>
+        /// <param name="tempFolder">The temp folder to use for storing assembly files</param>
         /// <param name="tracer">The tracer</param>
-        public SmartDetectorLoader(ITracer tracer)
+        public SmartDetectorLoader(string tempFolder, ITracer tracer)
```
Comment (EranGon) [inline src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs]: >string tempFolder [](start = 35, length = 17)

feels weird to inject this 
maybe send as another parameter to LoadSmartDetector?

File: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs (hunk 12 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs]: Since the assembly needs to be searched in the temp folder

---
In reply to: [186693107](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186693107) [](ancestors = 186693107)

File: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs (hunk 15 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs]: Considered it, but it means I will have to pass the tempFolder through the SmartDetectorRunner (and other classes). I think this is cleaner.

---
In reply to: [186699981](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186699981) [](ancestors = 186699981)

**Comment by liadtal** [comment]
>     /// Implementation of the <see cref="ISmartDetectorLoader"/> interface.

Just for you, I added documentation :-)

---
In reply to: [387361681](https://github.com/liadtal/Smart-Alerts/pull/150#issuecomment-387361681) [](ancestors = 387361681)

---
Refers to: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs:18 in 071e23e. [](commit_id = 071e23eade70d3d43f62fdbb56636a82a78b6d48, deletion_comment = False)

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs (hunk 6 above)
Comment (liadtal) [reply src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: You're right. Added an arguments class to handle this.

---
In reply to: [186687832](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186687832) [](ancestors = 186687832)

File: src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs (hunk 3 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/Extensions/StringExtensions.cs]: No, but its a waste of time to setup the CSharp scripting engine in these cases

---
In reply to: [186685334](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186685334) [](ancestors = 186685334)

File: src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs (hunk 4 above)
Comment (liadtal) [reply src/SDK/SmartDetectorsShared/SmartDetectorLoader/SmartDetectorLoader.cs]: If the file exists, something's very wrong. The folder name has a GUID in it, so it should be empty at this stage

---
In reply to: [186685897](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186685897) [](ancestors = 186685897)

File: src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs (hunk 8 above)
Comment (liadtal) [reply src/SmartDetectorsApplianceShared/ChildProcess/ChildProcessManager.cs]: I agree - done

---
In reply to: [186688696](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186688696) [](ancestors = 186688696)

File: src/SDK/MonitoringApplianceEmulator/App.xaml.cs [hunk 16]
```diff
@@ -36,9 +40,14 @@ public partial class App : Application
         /// <param name="e">A <see cref="T:System.Windows.StartupEventArgs" /> that contains the event data.</param>
         protected override void OnStartup(StartupEventArgs e)
         {
+            // Prepare temp folder
+            tempFolder = Path.Combine(Path.GetTempPath(), "MonitoringApplianceEmulatorTemp", Process.GetCurrentProcess().Id.ToString());
```
[the remark this answers is not in the document]
Comment (liadtal) [reply src/SDK/MonitoringApplianceEmulator/App.xaml.cs]: I changed it to use a GUID

---
In reply to: [186693901](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186693901) [](ancestors = 186693901)

File: src/SDK/MonitoringApplianceEmulator/App.xaml.cs [hunk 17]
```diff
@@ -36,9 +40,14 @@ public partial class App : Application
...
+            this.TryDeleteTempFolder();
+            Directory.CreateDirectory(tempFolder);
```
[the remark this answers is not in the document]
Comment (liadtal) [reply src/SDK/MonitoringApplianceEmulator/App.xaml.cs]: In the appliance, the temp folder is created by the parent process. The loader code (which is run by the child process), does not create this folder

---
In reply to: [186693932](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186693932) [](ancestors = 186693932)

File: test/SmartDetectorsSharedTests/SmartDetectorLoaderTests.cs (hunk 5 above)
Comment (liadtal) [reply test/SmartDetectorsSharedTests/SmartDetectorLoaderTests.cs]: The current unit tests pretty much cover all of the scenarios - loading multiple DLLs, with dependencies, etc. They needed to be adapted to the new temp folder, but I don't think additional tests are needed.

---
In reply to: [186687372](https://github.com/liadtal/Smart-Alerts/pull/150#discussion_r186687372) [](ancestors = 186687372)
